### ISSUE TITLE
Support for dark mode favicons

### ISSUE THREAD

**Issue by felipecrs** [facebook/docusaurus#4954]
## 🚀 Feature

A new option right beside the current `favicon` in the `docusaurus.config.js`:

```js
module.exports = {
  favicon: 'img/favicon_black.svg',
  faviconDark: 'img/favicon_white.svg',
```

This is important because some logos are only black or white, and the black version does not fit well in dark browsers, while the white version does not fit well in light theme in browsers. For example, the current GitHub (themed, changes when you change the system preference) vs speedtest.net favicon (which is not themed): 
![image](https://user-images.githubusercontent.com/29582865/121697383-6224f000-caa3-11eb-89d6-ef3a54563f6d.png)
![image](https://user-images.githubusercontent.com/29582865/121699769-99949c00-caa5-11eb-8ea1-263c30e74364.png)

There are some alternatives to that, such as [using an SVG as the icon and theming it dynamically](https://medium.com/swlh/are-you-using-svg-favicons-yet-a-guide-for-modern-browsers-836a6aace3df)

Actually, there is tons of guides on the internet on how to accomplish that in all different browsers.

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes.

## Motivation

My icon is either black or white, and my company's design guideline does not allow me to paint the icon.

## Pitch

I believe this is a cool feature that could be exposed by a simple configuration in the Docusaurus config, while any complexity can be handled by Docusaurus automatically under the hoods.

**Comment by jsamr** [comment]
@felipecrs Note that you can use a `style` tag in your svg, check-out this article https://css-tricks.com/dark-mode-favicons/

```xml
<svg width="100" height="100" xmlns="http://www.w3.org/2000/svg">
  <style>
    circle {
      fill: yellow;
      stroke: black;
      stroke-width: 3px;
    }
    @media (prefers-color-scheme: dark) {
      circle {
        fill: black;
        stroke: yellow;
      }
    }
  </style>
  <circle cx="50" cy="50" r="47"/>
</svg>
```

**Comment by slorber** [comment]
What I understand is that GitHub switch the markup on system colormode preference:

```html
<link rel="icon" class="js-site-favicon" type="image/svg+xml" href="https://github.githubassets.com/favicons/favicon.svg">
```

```html
<link rel="icon" class="js-site-favicon" type="image/svg+xml" href="https://github.githubassets.com/favicons/favicon-dark.svg">
```

I don't think it's the best way to do that for a static website as we have to hardcode one of the 2 icons in the markup at build time.

Having one icon that would have a mediaquery like @jsamr looks like a better solution to me, so I'm going to close this for now unless you have strong arguments for not using this approach.

In any case, you can implement this logic of updating the favicon markup according to system preferences yourself by adding a custom JS script to your site

**Comment by felipecrs** [comment]
> @felipecrs Note that you can use a `style` tag in your svg, check-out this article https://css-tricks.com/dark-mode-favicons/
> 
> ```
> <svg width="100" height="100" xmlns="http://www.w3.org/2000/svg">
>   <style>
>     circle {
>       fill: yellow;
>       stroke: black;
>       stroke-width: 3px;
>     }
>     @media (prefers-color-scheme: dark) {
>       circle {
>         fill: black;
>         stroke: yellow;
>       }
>     }
>   </style>
>   <circle cx="50" cy="50" r="47"/>
> </svg>
> ```

How do I do this with the website generated by Docusaurus?

**Comment by slorber** [comment]
> how do I do this with the website generated by Docusaurus?

Try creating an svg in `/static/img/favicon.svg` with that svg content and use `favicon: "/img/favicon.svg"`

Maybe we'll also need to set a `type="image/svg+xml"`, let me know if this does not work

**Comment by felipecrs** [comment]
It works like a charm, but sadly Google Chrome only updates the icon style if I manually reload the page. Firefox updates instantaneously, very cool.

**Comment by jsamr** [comment]
@felipecrs I haven't tried it with docusaurus config file, but usually we use the URL param trick: `favicon: "/img/favicon.svg?version=1"`

**Comment by felipecrs** [comment]
`favicon: '/img/favicon.svg'` works also

**Comment by rakleed** [comment]
@slorber there is a problem with the [option proposed by `@jsamr`](https://github.com/facebook/docusaurus/issues/4954#issuecomment-860258531) - if you open the site in one theme and then change the system theme, the icon will not change.

**Comment by felipecrs** [comment]
> It works like a charm, but sadly Google Chrome only updates the icon style if I manually reload the page. Firefox updates instantaneously, very cool.

This continues to be the case, 3+ years later.

But I'm reading some articles, and maybe something can be done about it:

https://phuoc.ng/collection/html-dom/change-the-favicon-dynamically-based-on-user-color-scheme-preference/

@slorber would you mind reviewing this? It states two different options to change the favicon based on user's theme, maybe something in Docusaurus can be done based on it.

**Comment by slorber** [comment]
Something worth mentioning: there are 2 themes:
- The one of your OS
- The one of your Docusaurus site (that may sync to your OS theme depending on params)

I think you want the favicon to sync to your OS theme (that controls the color of your browser tabs), not your Docusaurus site theme. So I think the media link solution may be the best option for that. 

Apparently, you can use it on link tags, but I think it doesn't work in Firefox?

https://phuoc.ng/collection/html-dom/change-the-favicon-dynamically-based-on-user-color-scheme-preference/#changing-favicon-dynamically-using-html-link-tags

Or you can use SVG favicon, but I think it doesn't work in Safari?

https://css-tricks.com/dark-mode-favicons/
https://caniuse.com/link-icon-svg

Note that GitHub probably uses JS instead of media. It likely works everywhere, but add an extra inlined JS script to all static pages, so it has a higher cost for little ROI. This also means that the favicon will not update automatically until you focus on the tab:

![CleanShot 2024-12-05 at 16 29 55](https://github.com/user-attachments/assets/0699a6dd-e873-4be0-8364-375c0cfbc9ff)

---

TLDR: there's no ideal solution. All 3 options have pros/cons. 

Depending on your icon's contrast, you may consider the "dark favicon" is a progressive enhancement, or not.

In any case, we provide the hooks to:
- insert an inlined JS script
- insert extra link tags
- reference an SVG file

So, you can already implement this on your own. Since choosing any of those options is opinionated, low ROI, low traction, I'd prefer to see successful userland implementations first before adding an opinionated feature in Docusaurus.

**Comment by mike-solomon** [comment]
Just in case this helps anyone else -- [I was able to get dark/light mode favicons working](https://github.com/moderneinc/moderne-docs/commit/db02839149a4dcde10a5ca6217279c6ba7f9f382) by updating the `headTags` array in the `docusaurus.config.ts` file:

```ts
{
  tagName: 'link',
  attributes: {
    rel: 'icon',
    href: '/img/favicon.ico',
    type: 'image/x-icon',
    sizes: '32x32',
    media: '(prefers-color-scheme: light)',
  },
},
{
  tagName: 'link',
  attributes: {
    rel: 'icon',
    href: '/img/favicon_dark.ico',
    type: 'image/x-icon',
    sizes: '32x32',
    media: '(prefers-color-scheme: dark)',
  },
},
```

No idea if that's a bad idea or not, but it's working for us. Please note that this reflects the OS theme (so if your OS is set to dark mode it will use the dark mode icon and vice versa).

**Comment by slorber** [comment]
Safari 26 beta adds support for SVG favicons: https://webkit.org/blog/16993/news-from-wwdc25-web-technology-coming-this-fall-in-safari-26-beta/
